command/command.go

```diff
@@ -0,0 +1,126 @@
+package command
+
+import (
+	"bufio"
+	"io"
+	"os"
+	"os/exec"
+	"sync"
+)
+
+// ExternalCommand has the properties for invoking exec.Command
+type ExternalCommand struct {
+	Path string
+	Arg  []string
+	Env  []string
+}
+
+type ProgramState struct {
+	Command     *exec.Cmd
+	stdoutPipe  io.ReadCloser
+	stderrPipe  io.ReadCloser
+	stdout      []byte
+	stderr      []byte
+	stdoutmutex sync.Mutex
+	stderrmutex sync.Mutex
+}
+
+// ExecuteProgram invokes an external program
+func ExecuteProgram(command ExternalCommand, errorHandler func(*ProgramState, error) error) (*ProgramState, error) {
+	var err error
+	programState := &ProgramState{}
+	cmd := exec.Command(command.Path, command.Arg...)
+	cmd.Env = append(os.Environ(), command.Env...)
+	programState.Command = cmd
+
+	programState.stdoutPipe, err = cmd.StdoutPipe()
+	if err != nil {
+		return programState, err
+	}
+	programState.stderrPipe, err = cmd.StderrPipe()
+	if err != nil {
+		return programState, err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return programState, err
+	}
+
+	go CloseProgram(programState, errorHandler)
+
+	return programState, err
+}
+
+func CloseProgram(programState *ProgramState, errorHandler func(*ProgramState, error) error) {
+
+	var waitGroup sync.WaitGroup
+	waitGroup.Add(1)
+
+	go func() {
+		errStdout := programState.StdoutListner()
+		if errStdout != nil {
+			errorHandler(programState, errStdout)
```

Comment:
errorHandler is only for the case of program exiting. Documentation says that this error can _only_ be eof, so just ignore it, and stop reading from this pipe.